### ISSUE TITLE
Add continuous tracking of vessels by rule

### ISSUE THREAD

**Issue by Keniamin** [jrossignol/ContractConfigurator#469]
Hello! Now VesselParameterGroup has only two options to track a vessel: define and defineList. First one works continuously, but defines only one vessel. Second one defines many vessels, but only at the end of contract. So there is no way to determine, during the contract, all vessels that match some rule.

Use case: a vessel with two kerbals, X and Y, aboard that will decouple and became two vessels with one kerbal at each. If I want to avoid destruction of any of them I have to determine two VesselParameterGroup, one for "has X aboard", another for "has Y aboard", each one would define a vessel and then I determine VesselNotDestroyed parameter with both of them. Moreover, if X then performs EVA and goes to third vessel, it will not be tracked at all (as I understand, though not checked it), because both "define" already tracking something.

Solution: to introduce parameter like "defineListContinuously", which would works like defineList, but works continuously (at any state change of VesselParameterGroup, like simple define) instead of on completion. For the use case above it allows to determine one VesselParameterGroup with condition "any of X, Y aboard" and "defineListContinuously = SaveVessels", and then VesselNotDestroyed with ITERATOR of @/SaveVessels. Starting vessel will be in list (has both X and Y), on decoupling (precisely, on first selection of decoupled part) second vessel will also be added to the list, on X changing vessel third one became matching so became tracked. Moreover, on further selection of original vessel it would not matches group anymore, so can be deleted from the list and thus lost tracking. Briefly, it looks like quite simple but powerful solution.

**Comment by jrossignol** [comment]
I'm not 100% clear on the problem that you're trying to solve with this, but there are a few issues (due to the way that KSP and Contract Configurator work):

1. All expressions are executed and contract parse time - so there can't be any ITERATOR nodes that have stuff added to them dynamically.
1. Undocking actually only creates one new vessel (and it's half-random which of the two parts of the vessel keeps the old id).
1. Some of the vessel checks only work on the active vessel.  Rather, all vessel checks can track across any vessel, but a lot of the checks don't happen except on active vessels.  What you're looking for could require active checking of all vessels at all times (very performance intensive).

If you can describe the contract you're trying to create, I might be able to suggest an alternative within the existing framework that could work (or possibly an enhancement of smaller scope).

**Comment by Keniamin** [comment]
> Undocking actually only creates one new vessel

Yes , it's clear. I saw no problem with this behaviour for described variant.


> What you're looking for could require active checking of all vessels at all times

No-no-no, of course not. I supposed updating of the list might be performed near the updating of simple define (i.e., in the OnParameterStateChange function in VesselParameterGroup.cs). It would add/delete vessels from the list only on it's selection (or while it is active) and thus work only for vessels, that was selected at least once during the contract (so fired this event). But for my case that's enough.


But yes,

> there can't be any ITERATOR nodes that have stuff added to them dynamically

is a real problem. I think this breaks my solution completely.


> If you can describe the contract you're trying to create

Right now I'm trying to solve problem from another my issue #470 :) That is, avoid deaths of some specified kerbals. I tried to define a hidden VesselParameterGroup with condition "have any of passengers aboard" (implemented as Any of HasCrew with ITERATOR), and define the vessel that matches as SafeVessel. And then VesselNotDestroyed on it. It works perfectly if all passengers are in the same vessel, but if i put them to vessel, then on EVA transfer one of them to another vessel, and then destroys this second vessel with kerbal aboard – game still tracks the first one, so does not recognize this case as failure. And my thought was to add ability to track both this vessels (because both of them at least at one moment was in situation "active and mathes group", so OnParameterStateChange was fired and vessel got tracking).
